packages/hw-transport-node-hid/src/listenDevices.js

```diff
@@ -27,37 +37,93 @@ export default (
 
   let lastDevices = getFlatDevices();
 
-  const checkDevices = () => {
+  const poll = (type, attempt = 1) => {
+    let changeFound = false;
+
     if (!listenDevicesPollingSkip()) {
+      debug(`Polling for ${type} [attempt ${attempt}/${MAX_ATTEMPTS}]`);
+
       const currentDevices = getFlatDevices();
 
-      const newDevices = currentDevices.filter(d => !lastDevices.includes(d));
-      const removeDevices = lastDevices.filter(
-        d => !currentDevices.includes(d)
-      );
+      if (type === "add") {
+        const newDevices = currentDevices.filter(d => !lastDevices.includes(d));
+
+        if (newDevices.length > 0) {
+          debug("New device found:", newDevices);
+
+          listDevices = getDevices();
+          events.emit("add", getDeviceByPaths(newDevices));
 
-      if (newDevices.length > 0) {
-        listDevices = getDevices();
-        events.emit("add", getDeviceByPaths(newDevices));
+          changeFound = true;
+        } else {
+          debug("No new device found");
+        }
       }
 
-      if (removeDevices.length > 0) {
-        events.emit("remove", getDeviceByPaths(removeDevices));
-        listDevices = listDevices.filter(
-          d => !removeDevices.includes(flatDevice(d))
+      if (type === "remove") {
+        const removeDevices = lastDevices.filter(
+          d => !currentDevices.includes(d)
         );
+
+        if (removeDevices.length > 0) {
+          debug("Removed device found:", removeDevices);
+
+          events.emit("remove", getDeviceByPaths(removeDevices));
+          listDevices = listDevices.filter(
+            d => !removeDevices.includes(flatDevice(d))
+          );
+
+          changeFound = true;
+        } else {
+          debug("No removed device found");
+        }
+      }
+
+      if (changeFound) {
+        lastDevices = currentDevices;
+      } else {
+        if (attempt < MAX_ATTEMPTS) {
+          const newDelay = delay * attempt;
+
+          debug(`Repolling ${type} in ${newDelay}ms`);
+
+          setTimeout(() => {
+            poll(type, attempt + 1);
+          }, newDelay);
+        } else {
+          debug(`Giving up after ${attempt} attempts`);
+        }
       }
+    } else {
+      debug(`Polling skipped, retrying in ${delay}ms`);
 
-      lastDevices = currentDevices;
+      setTimeout(() => {
+        poll(type);
+      }, delay);
     }
-    setTimeout(checkDevices, delay);
   };
 
-  timeoutDetection = setTimeout(checkDevices, delay);
+  debug("Starting to monitor USB for ledger devices");
+  usbDetect.startMonitoring();
+
+  // Detect add
+  usbDetect.on(`add:${VENDOR_ID}`, device => {
```

Comment:
https://github.com/LedgerHQ/ledgerjs/blob/master/packages/hw-transport-node-hid/src/isLedgerDevice.js

maybe we should filter like it was here?

Reply:
This is a pre-filter, taking advantage of the vendor ID filtering offered by `usb-detection`, our `isLedgerDevice` is used by `getDevices` afterwards

Reply:
👌